### ISSUE TITLE
password hash injection

### ISSUE THREAD

**Issue by loblik** [grw1983/netio4#1]
Hi,

trying to do the same thing. Have you made any progress on this?

**Comment by grw1983** [comment]
I was not able to decipher the password.
What I did so far:
- compile a custom bootloader and flash it to the EEPROM -> boots
- read out most of the eMMC (somewhat useless because my eMMC is likely damaged)
- create a custom dtb and compile a linux kernel for the NETIO4 -> boots

With my damaged eMMC I was also not able to run the binaries which would enable me to decrypt the Koukaam update files (so I could rebuild the damaged system) - this is pretty well closed with two or more cipher stages combined.

If you are interested I can share my builds/logs.

**Comment by loblik** [comment]
What U-Boot sources you used for custom build?

I'm trying to run upstream version of U-Boot on this. I'm close to have it working, but not sure if I will be able squeeze everything in bootloader partition on SPI flash. I went with three stage boot scheme (TPL->SPL->U-Boot), since loading U-Boot using SPL running from SPI flash seems a bit slow.

I would like to build Armbian or OpenWRT for the board in the end. Once I get some ramdisk running I could make a backup of the partitions and send it to you. If that helps to recover your system.

**Comment by grw1983** [comment]
I'm using https://github.com/u-boot/u-boot.git (my HEAD is at "https://github.com/u-boot/u-boot/commit/da2e3196e4dc28298b58a018ace07f85eecd1652") - I will ad my config to the repo.

I will also upload my kernel build for your convenience.

**Comment by loblik** [comment]
Your U-Boot configuration is simpler and boots faster, cool. It runs fine on my board, I guess they are the same revision.

Not sure why it was so slow when I tried to load U-Boot proper this way. Probably I had something else wrong. And it's also close to the size limit of bootloader SPI flash partition, only about ~10k over.  It was easy to get it under the limit by dropping a few config options.

Also the ramdisks runs fine. If you could make a build with eMMC debug lines disabled I could create that eMMC backup. Reading is quite slow probably because these debug lines (even if I suppress console prints with dmesg -n 1).

**Comment by grw1983** [comment]
As you wish. ;-D

**Comment by loblik** [comment]
I mean, you don't have to do that. I will be playing with OpenWRT anyway later. I just thought you might be interested in these dumps, since your files got corrupted.

**Comment by grw1983** [comment]
I was just kidding. The new kernel is already commited to the repo.

**Comment by loblik** [comment]
Thanks. This is zstd compressed full eMMC image  https://cloud.loebl.cz/s/oZaBW9xx5BdPZx3 .

sha256 sum:

`b8d25b88d08326c835a557dc1b170aa6405ee6ee54a2b42b954756200378bc42  mmcblk0.img`

Have you been able to get into the running system? Or is it not really bootable after data corruption? I'm mainly interested where the MAC addresses and radio calibration data are stored. I guess radio parameters should be in factory partition on SPI flash, but there is also partition called calibration on eMMC. Have to check if it's a backup or something.

I played a bit with partitions from the ramdisk and it seems is not so easy to mess with shadow file. But there are other executable scripts directly in cramfs on mmcblk0p3, so I  there will be a way I guess to get root. However it seems they tried to make it a bit harder. I did not expect the firmware be that closed, considering how old it is.

Edit: Ok, mmcblk0p8 (calibration) is full of 0xff and mmcblk0p6 (extensions) is full of zeros. So it's probably in SPI.

Edit2:

Mac addresses and calibration data are really stored in factory partition of SPI flash. Looking at the hexdump and existing devicetrees for similar boards, layout of factory partition seems pretty standard.

Hardware-wise: 

There is MCP7940M real time clock on I2C, 74HC4094D shift register controlling 4 RG LEDs, micro switches (buttons) and buzzer are probably on GPIOs.

Relays switching the sockets are controlled from dedicated MCU connected to the SOC using I2C or SPI. So the communication must be sniffed on stock firmware.

**Comment by grw1983** [comment]
- No chance to get into the running system - my eMMC is damaged
- Yes. It's very well closed. I tried to decrypt the original FW update files by manipulating the update script. But as my system is damaged the involved update scripts/executables are damaged, also  (I actually tried to solder a new eMMC, but I doubt this will work and is a work in progress at the moment anyway).
- Yes. the MAC is in the NOR Flash - makes sense actually.

Thank you for the dump - If I'm able to repair my system, the dump will come very handy!

Edit:
Koukaam - as it seems - came from the IP-Camera Recorder business (the ipcorder partition kinda clued me in). It looks like they just modified this system and crammed it onto a network power strip. That's probably why it is very well closed.

**Comment by loblik** [comment]
So you think you have faulty eMMC? Actually with OpenWRT build you could just swap the NOR flash for bigger one, like W25Q128, and run the board without eMMC. In case reflow will not work. Most ramips based routers are running happily like this I believe.

**Comment by grw1983** [comment]
Yes, I had lots of `blk_update_request: I/O error, dev mmcblk0, sector xxx op 0x0:(READ)`. That's the reason I started debugging the SDHCI driver and playing with various quirk settings. At the time I thought that maybe the eMMC is not 100% adhering to the standard.

The fact that the same kernel build obviously works fine with your hardware is the final proof that my hardware is faulty.
